### PR TITLE
Refactor ML code and add support for multiple KMeans models.

### PR DESCRIPTION
**Pull request by vkalintiris**
A number of changes are introduced in this PR to simplify the ML code and add support for multiple KMeans models.

Notable changes:

- Use a thread-safe queue to submit and process training requests
- Keep track of the result status of every step, so that we can understand why/when a dimension is not trained, etc.
- Add configuration option to support multiple KMeans models.

##### Test Plan

- CI jobs
- Tested by running multiple VMs for long periods of time under different configuration settings.

### CODE REVIEW

**Comment by andrewm4894** [comment]
Some charts after about 18 hours or so. 

cpu profile of netdata looks better on new branch:

image
image

mem looks similar:

image

anomaly rates looking flat for `ml-demo-ml-enabled` - i will need to look into what's going on here:

image

oh i can see anomaly rate looks flat on that node as its misbehsaving such that supresses rates:

image

So limited ways to compare anomaly rates across ml-demo-ml-enabled and the other nodes.

I think i will leave all this running for a while so we can see performance characteristics of it all. 

Then i will turn off all my docker containers and extra stuff to try get ml-demo-ml-enabled running normal from ml point of view and hopefully be able then to compare the anomaly rates across nodes too.

**Comment by andrewm4894** [comment]
just cross posting this discussion as background https://github.com/netdata/netdata/discussions/12763

File: ml/Config.cc [hunk 1]
```diff
@@ -84,10 +84,17 @@ void Config::readMLConfig(void) {
 
     Cfg.EnableAnomalyDetection = EnableAnomalyDetection;
 
+ #if 1
     Cfg.MaxTrainSamples = MaxTrainSamples;
     Cfg.MinTrainSamples = MinTrainSamples;
     Cfg.TrainEvery = TrainEvery;
-    Cfg.NumModelsToUse = NumModelsToUse;
+    Cfg.NumModelsToUse = 1;
```
Comment (andrewm4894) [inline ml/Config.cc:91]: @vkalintiris just checking this is not hard coding `NumModelsToUse` to 1 or anything?

Comment (andrewm4894) [reply ml/Config.cc]: cool - i will drop and recreate the test nodes when ready

**Comment by andrewm4894** [comment]
updating with some ml results 

[link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/anomalies#highlight_after=1670587128000&highlight_before=1670587270000&after=1670586884000&before=1670587484000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&tocSearch=&metrics_correlation=false&appliedFilteredIds=d5ae15cd-b12f-48b5-8147-d74a37fc4fb4%2C710ef39b-60d6-47ee-b93e-1b6a8766961a%2C96465dc4-1b3f-4dd7-926f-f0545c3475a5%2C9de54c34-8c45-433b-8f07-a6062159f1ee%2C024425f1-15fa-4e26-b52c-b97e2bf3ac43%2C5f36dd02-8063-415d-ad6b-34262a1bf72f%2C6f2b28a3-c912-4916-91ba-a255e7320bab%2C1c15d7c5-3e71-4ada-b666-ee03a09dc34d%2C3d4fe579-6b97-49e4-afa3-a03c1ce473c7&liveNodeIds=d5ae15cd-b12f-48b5-8147-d74a37fc4fb4%2C710ef39b-60d6-47ee-b93e-1b6a8766961a%2C96465dc4-1b3f-4dd7-926f-f0545c3475a5%2C9de54c34-8c45-433b-8f07-a6062159f1ee%2C024425f1-15fa-4e26-b52c-b97e2bf3ac43%2C5f36dd02-8063-415d-ad6b-34262a1bf72f%2C6f2b28a3-c912-4916-91ba-a255e7320bab%2C1c15d7c5-3e71-4ada-b666-ee03a09dc34d%2C3d4fe579-6b97-49e4-afa3-a03c1ce473c7&staleNodeIds=&offlineNodeIds=&filteredNodesSearchParams=&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_anomaly_detection_submenu_anomaly_rate) to highlighted nodes. 

## Load test to trigger a true anomaly on all nodes

goal was to hit all nodes with a k6 load test and so simulate a true unknown/unseen anomaly on these nodes and see if node anomaly rates behave as expected. 

this was the profile of the k6 test: 
image

corresponding anomaly advisor [link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/anomalies#highlight_after=1670587128000&highlight_before=1670587553000&after=1670586734000&before=1670587634000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&tocSearch=&metrics_correlation=false&appliedFilteredIds=d5ae15cd-b12f-48b5-8147-d74a37fc4fb4%2C710ef39b-60d6-47ee-b93e-1b6a8766961a%2C96465dc4-1b3f-4dd7-926f-f0545c3475a5%2C9de54c34-8c45-433b-8f07-a6062159f1ee%2C024425f1-15fa-4e26-b52c-b97e2bf3ac43%2C5f36dd02-8063-415d-ad6b-34262a1bf72f%2C6f2b28a3-c912-4916-91ba-a255e7320bab%2C1c15d7c5-3e71-4ada-b666-ee03a09dc34d%2C3d4fe579-6b97-49e4-afa3-a03c1ce473c7&liveNodeIds=d5ae15cd-b12f-48b5-8147-d74a37fc4fb4%2C710ef39b-60d6-47ee-b93e-1b6a8766961a%2C96465dc4-1b3f-4dd7-926f-f0545c3475a5%2C9de54c34-8c45-433b-8f07-a6062159f1ee%2C024425f1-15fa-4e26-b52c-b97e2bf3ac43%2C5f36dd02-8063-415d-ad6b-34262a1bf72f%2C6f2b28a3-c912-4916-91ba-a255e7320bab%2C1c15d7c5-3e71-4ada-b666-ee03a09dc34d%2C3d4fe579-6b97-49e4-afa3-a03c1ce473c7&staleNodeIds=&offlineNodeIds=&filteredNodesSearchParams=&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_anomaly_detection_submenu_anomaly_rate)

as expected all nodes see a jump in node anomaly rate so this is good. 

image

as might be expected the node with just 1 model tends to have higher jump in anomaly rate than that of node with 24 models (probably not all trained yet actually since nodes only maybe 18 hours old or so). but the difference is not massive and thats good - for a real unseen type anomaly like this we want the node anomaly rate to properly react:

image

if you look at top dimension then we see what we want - anomaly rate over all nodes tends to jump to 100% for all nodes for the most affected dims - that's great as means that that particular dim immediately reacts on all nodes for all configs. So on the hardest node with 24 models this means that it really was considered anomalous - which makes sense as this was completely new behaviour for the nodes. 

image

we can see this clearly on the corresponding chart of overview screen with the anomaly rate chart expanded and group by node with the relevant dimension selected:

image

## General observations

If we look at the overall node anomaly rates over time we see some expected characteristics

1. after the nodes are created the node anomaly rates are a bit noise for a couple of hours while enough data builds up for proper models to be trained.
image

2. as expected, if you pick any random window of time, the nodes that have more models tend to have a much more depressed overall node anomaly rate. this effect seems to kick in quick - even comparing node with 1 model to node with 3 models you can clearly see it 
image
image

3. this means that for nodes with lots of models you tend to see a much smoother anomaly rate in general with more clear spikes when something anomalous looking occurs. while overall node level anomaly rate itself is just generally lower. for example below is node with same workloads but one only uses 1 model and the other uses 24. You can see that the red line is sort of more like a less noisy version of the purple line. 
image

## Conclusion

In terms of the ML side of things this all looks good to me. Implications are that anomaly rates will tend to be less noisy, generally a bit lover in terms of node anomaly rate, but should be more reliable in that when a spike or bump does occur somewhere it's more likely on average to be a true anomaly since more models have been checked as part of the inference process.

@vkalintiris we can not look at the performance characteristics of the nodes in the ml-demo space to see what we think about that side of things.

**Comment by andrewm4894** [comment]
internal video showing some other additional tests: https://drive.google.com/file/d/17nW5weEYPAvk-99O53dsILg5_nQBUZpL/view?usp=share_link

**Comment by vkalintiris** [comment]
> @vkalintiris we can not look at the performance characteristics of the nodes in the ml-demo space to see what we think about that side of things.

AFAICT, it looks good. In fact, it looks way too good when comparing 1 vs 24 models and this makes me a little bit sceptical. I want to run one example locally and see how much we can stretch the amount of models that we can keep around before performance deteriorates.

**Comment by andrewm4894** [comment]
Yeah was expecting to see a bigger difference too in terms of the overhead

**Comment by andrewm4894** [comment]
I wonder if a way to add a hidden param to force checking of all models to get a worst case bounds

**Comment by vkalintiris** [comment]
> I wonder if a way to add a hidden param to force checking of all models to get a worst case bounds

Added [ml] option "consume all models = false". Change that to true and we will check every single model. Didn't see any difference in performance with up to 10 models.

Truth be told, we were doing a couple things quite inefficiently prior to this PR. These inefficiencies have been fixed with this PR, hence why performance remains more or less the same even when using multiple models. At least that's my guess.

**Comment by andrewm4894** [comment]
[adding some more test nodes](https://github.com/netdata/ml/pull/77/files) to get upper bound on overhead, once deployed will check in ~24 hours and report findings in here.

**Comment by andrewm4894** [comment]
Updating with some `-all` vs not all nodes.

[link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/overview#after=1670873447000&before=1670918432000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&tocSearch=&metrics_correlation=false&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_apps&appliedFilteredIds=9e053459-a334-4785-b8b2-6106cd5a5d27%2C1876eb46-4299-449a-afe6-62ccecd42e87&liveNodeIds=9e053459-a334-4785-b8b2-6106cd5a5d27%2C1876eb46-4299-449a-afe6-62ccecd42e87&staleNodeIds=&offlineNodeIds=&filteredNodesSearchParams=)

looks like about 20% extra cpu in this worse case for my proposed newconf default which is great.

image

prediction usage seems maybe 10% more on average which is also good: 

image

training same as expected

image

looking at the train hourly current default models even 24 models is only really a little higher everywhere.

[link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/overview#appliedFilteredIds=e0b8b19a-b15d-46cf-ac30-5f8dba346009%2Ca7160189-d66e-44fd-b8ac-476f6fbed450%2C2989cc5b-0e1e-4ba9-9607-6e8342c79e33%2Ccd868bb1-3a55-402f-a0e7-fa05267213a3%2C6c8dc03e-16ae-41ae-85e5-74a2342eb0ee%2C2e9f2813-c9ef-4f6e-9678-210a5aa50f07%2Ca6166e53-589d-44ee-ba13-e1b8db3adbfb&liveNodeIds=e0b8b19a-b15d-46cf-ac30-5f8dba346009%2Ca7160189-d66e-44fd-b8ac-476f6fbed450%2C2989cc5b-0e1e-4ba9-9607-6e8342c79e33%2Ccd868bb1-3a55-402f-a0e7-fa05267213a3%2C6c8dc03e-16ae-41ae-85e5-74a2342eb0ee%2C2e9f2813-c9ef-4f6e-9678-210a5aa50f07%2Ca6166e53-589d-44ee-ba13-e1b8db3adbfb&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_apps&after=1670873447000&before=1670918432000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&tocSearch=&metrics_correlation=false&staleNodeIds=&offlineNodeIds=&filteredNodesSearchParams=)

image

So all in all this is looking great to me in terms of worse case bounds relative change in overhead.

**Comment by andrewm4894** [comment]
@vkalintiris have a look at links above yourself. 

the `-all` are versions of same node conf but with all models being forced for checking.

image

**Comment by andrewm4894** [comment]
Next steps i can see

1. rename `anomaly_detection.predictions` to `anomaly_detection.dimensions` so that it can serve the AA tab chart number 2.
2. apart from `anomaly_detection.dimensions`, `anomaly_detection.anomaly_rate`, `anomaly_detection.detector_events` move all other charts into `ml` section under `netdata monitoring`

Then we see if we ready to start wider review process for merging?

**Comment by vkalintiris** [comment]
> Next steps i can see
> 
> 1. rename `anomaly_detection.predictions` to `anomaly_detection.dimensions` so that it can serve the AA tab chart number 2.
> 2. apart from `anomaly_detection.dimensions`, `anomaly_detection.anomaly_rate`, `anomaly_detection.detector_events` move all other charts into `ml` section under `netdata monitoring`
> 
> Then we see if we ready to start wider review process for merging?
> 
> Then, in new follow-on PR's (?)
> 
> 1. model persistence to db.
> 2. extending to cover additional update every of 2,5,10 (or arbitrary - whatever is easiest implementation).

Sounds like a plan! I'm going to need some help with updating the docs as well.

**Comment by andrewm4894** [comment]
This all looks good to me. I will make the relevant doc changes as part of this PR so it's all self contained. 

Think we ready to finalize this PR.

File: ml/Config.cc [hunk 2]
```diff
@@ -31,7 +31,7 @@ void Config::readMLConfig(void) {
     unsigned MaxTrainSamples = config_get_number(ConfigSectionML, "maximum num samples to train", 4 * 3600);
     unsigned MinTrainSamples = config_get_number(ConfigSectionML, "minimum num samples to train", 1 * 900);
     unsigned TrainEvery = config_get_number(ConfigSectionML, "train every", 1 * 3600);
-    unsigned NumModelsToUse = config_get_number(ConfigSectionML, "number of models per dimension", 1 * 24);
+    unsigned NumModelsToUse = config_get_number(ConfigSectionML, "number of models per dimension", 1);
```
Comment (andrewm4894) [inline ml/Config.cc:34]: makes sense to leave this like this (defaulting to just 1 model) and i will make a follow on PR to change the defaults on a few weeks after i dogfood my new proposed defaults a bit more first.

**Comment by andrewm4894** [comment]
image
image
image

File: ml/Dimension.cc [hunk 3]
```diff
@@ -144,30 +266,80 @@ bool Dimension::predict(CalculatedNumber Value, bool Exists) {
...
+
     std::unique_lock<std::mutex> Lock(Mutex, std::defer_lock);
     if (!Lock.try_lock()) {
-        AnomalyBit = false;
         return false;
     }
 
+    // Mark the metric time as variable if we received different values
+    if (!SameValue)
+        MT = MetricType::Variable;
```
Comment (thiagoftsm) [inline ml/Dimension.cc:280]: Hello @vkalintiris ,

I am on vacation, so I am only giving to you this coverity scan report:

```sh
** CID 432006:  Concurrent data access violations  (MISSING_LOCK)
/ml/Dimension.cc: 280 in ml::Dimension::predict(long, double, bool)()
274         if (!Lock.try_lock()) {
275             return false;
276         }
277     
278         // Mark the metric time as variable if we received different values
279         if (!SameValue)
>>>     CID 432006:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "this->MT" without holding lock "ml::Dimension.Mutex". Elsewhere, "ml::Dimension.MT" is accessed with "Dimension.Mutex" held 2 out of 3 times (1 of these accesses strongly imply that it is necessary).
280             MT = MetricType::Variable;
281     
282         // Decide if the dimension needs to be scheduled for training
283         scheduleForTraining(CurrT);
284     
285         // Nothing to do if we don't have a model
```

The PR is running as expected, but I did not have time today to review the code.

Best regards!

Comment (vkalintiris) [reply ml/Dimension.cc:280]: Don't think this specific report is accurate. We reach LOC 280 only if we managed to hold the lock earlier at LOC 274.

**Comment by andrewm4894** [comment]
@vkalintiris i just noticed that we dont have these "worker ML" charts on this branch vs master - are they superseded by some in the ML section or worth adding back in? The seem to have cpu % utilization specific to the ML threads that seemed a little easier for me to understand but not sure if relevant or not with the changes in this PR?

image

**Comment by vkalintiris** [comment]
> @vkalintiris i just noticed that we dont have these "worker ML" charts on this branch vs master - are they superseded by some in the ML section or worth adding back in? The seem to have cpu % utilization specific to the ML threads that seemed a little easier for me to understand but not sure if relevant or not with the changes in this PR?
> 
> image

These are now under "Netdata Monitoring"."ml". We will re-add worker charst in a follow up PR.

**Review by andrewm4894 — DISMISSED** [review]

**Comment by andrewm4894** [comment]
@vkalintiris one thing we might need to think about a little in terms of maybe handling or thinking about edge cases. 

I actually hit 3 nodes here with a real anomaly, which as expected triggered a jump in number of ml models used for prediction on the 2 nodes with saved models:

[link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/overview#metrics_correlation=false&after=1671451686000&before=1671451855000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&6045dc92-0a0c-48d8-960c-d2e0f31aa95d--selectedNodeIds=8ca367d1-e410-4789-870b-8df0388e48e4%2C13b68eec-7700-4be8-abdd-a89f15f4bead%2C61092b71-6df3-4233-90b2-043a7a0b9472&modalParams=&highlight_after=1671451793000&highlight_before=1671451817000&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-selectedNodeIds=8ca367d1-e410-4789-870b-8df0388e48e4&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-selected=processes&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-isPolling=true&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-pollingInterval=2&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_netdata_submenu_ml)
image

Then you can see a spike in cpu usage by netdata as a result:

[link](https://app.netdata.cloud/spaces/ml-demo-space/rooms/all-nodes/overview#metrics_correlation=false&after=1671451686000&before=1671451855000&utc=Europe%2FLondon&offset=%2B0&timezoneName=Edinburgh%2C%20London&modal=&modalTab=&6045dc92-0a0c-48d8-960c-d2e0f31aa95d--selectedNodeIds=8ca367d1-e410-4789-870b-8df0388e48e4%2C13b68eec-7700-4be8-abdd-a89f15f4bead%2C61092b71-6df3-4233-90b2-043a7a0b9472&modalParams=&highlight_after=1671451793000&highlight_before=1671451817000&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-selectedNodeIds=8ca367d1-e410-4789-870b-8df0388e48e4&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-selected=processes&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-isPolling=true&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-fn-pollingInterval=2&6045dc92-0a0c-48d8-960c-d2e0f31aa95d-overview-chartName=menu_apps_submenu_cpu)
image

So when an anomaly hits is when the ML will be checking most models and so will tend to have a temp jump in overhead from this. I'm not quite sure what we can do about this apart from in some way trying to cap the cpu usage or have netdata somehow throttle itself in this scenario but its a very tricky one as not really any obvious way around it that i can see. So i think it more so just a new characteristic that we need to be aware of and maybe think about over time if anything we can or should try do about it.

File: collectors/profile.plugin/README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+You can configure this plugin with the following options:
```
Comment (thiagoftsm) [inline collectors/profile.plugin/README.md:1]: Where can the user configure it? Please, take a look in the first paragraph of [apps.plugin](https://github.com/netdata/netdata/tree/master/collectors/apps.plugin#configuration) readme.

File: collectors/profile.plugin/README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+    - `profplug charts = 500`: number of charts to create
```
Comment (thiagoftsm) [inline collectors/profile.plugin/README.md:2]: If somehow this configuration is inside a `/etc/netdata/profile.d.conf`, I suggest to remove the `profplug` from the names.

On the other hand, if this is inside `/etc/netdata/netdata.conf`,  I suggest to move for an external configuration file.

File: collectors/profile.plugin/plugin_profile.cc [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+static std::vector<std::shared_ptr<Chart>> createCharts(size_t NumCharts, unsigned NumDimsPerChart, time_t UpdateEvery) {
+    std::vector<std::shared_ptr<Chart>> Charts;
+    Charts.reserve(NumCharts);
+
+    constexpr size_t Len = 1024;
+    char Buf[Len];
+
+    for (size_t ChartIdx = 0; ChartIdx != NumCharts; ChartIdx++) {
+        snprintfz(Buf, Len, "chart_%zu", ChartIdx + 1);
```
Comment (thiagoftsm) [inline collectors/profile.plugin/plugin_profile.cc:102]: In normal situation when we call this function, instead to give as argument the buffer size, we give a smaller size as you can see [here](https://github.com/netdata/netdata/blob/76d93161cd041774e95a23eac00057437a7f07e6/collectors/apps.plugin/apps_plugin.c#L1136-L1137).

Of course with the given size, I do not expect issues. I am only calling attention for our pattern.

File: collectors/profile.plugin/plugin_profile.cc [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+void *profile_main(void *ptr)
+{
+    netdata_thread_cleanup_push(profile_main_cleanup, ptr);
+
+    size_t NumCharts = config_get_number(CONFIG_SECTION_GLOBAL, "profplug charts", 1);
+    size_t NumDimsPerChart = config_get_number(CONFIG_SECTION_GLOBAL, "profplug dimensions", 1);
+    size_t SecondsToFill = config_get_number(CONFIG_SECTION_GLOBAL, "profplug seconds to fill", 0);
+    time_t UpdateEvery = 1;
```
Comment (thiagoftsm) [inline collectors/profile.plugin/plugin_profile.cc:139]: Why are we not giving possibility to change `UpdateEvery`?

File: database/rrdset.c [hunk 8]
```diff
@@ -1332,8 +1339,11 @@ static inline size_t rrdset_done_interpolate(
                     break;
             }
 
+            time_t curr_t = (time_t)(next_store_ut / USEC_PER_SEC);
```
Comment (thiagoftsm) [inline database/rrdset.c:1342]: This is a small one, only to think. We are using `time_t` and we created `ml_host_t `, so we are using the suffix `_t` to define types, so I suggest to sue a different name for `curr_t` avoiding confusion between data type and variable names.

File: ml/ADCharts.cc [hunk 9]
```diff
@@ -3,55 +3,199 @@
...
-
-        NumTotalDimensionsRD = rrddim_add(RS, "total", NULL,
-                1, 1, RRD_ALGORITHM_ABSOLUTE);
-        NumTrainedDimensionsRD = rrddim_add(RS, "trained", NULL,
-                1, 1, RRD_ALGORITHM_ABSOLUTE);
-        NumNormalDimensionsRD = rrddim_add(RS, "normal", NULL,
-                1, 1, RRD_ALGORITHM_ABSOLUTE);
-        NumAnomalousDimensionsRD = rrddim_add(RS, "anomalous", NULL,
-                1, 1, RRD_ALGORITHM_ABSOLUTE);
+#define NETDATA_ML_CHART_PRIO_PREDICTION_USAGE        890004
```
Comment (thiagoftsm) [inline ml/ADCharts.cc:16]: Considering the values I am seeing here, I suspect that these values were generated based  in values already present inside [this file](https://github.com/netdata/netdata/blob/master/collectors/all.h), if this is true, I suggest one of the following options:

- Define these values inside the file that gave the idea.
- Comment something inside the original file.

I believe that in both situation we will avoid future conflicts.

File: ml/Dimension.cc (hunk 3 above)
Comment (thiagoftsm) [reply ml/Dimension.cc:280]: When I reported initially I did not review code, now that I am reviewing, I agree with explanation so I am marking as resolved.

**Comment by thiagoftsm** [comment]
Hello @vkalintiris ,

I asked for some minor changes considering the codes we have for other collectors, but I have only one additional doubt: I am years without  to develop with `C++`, but when I read our code I missed a lot `try .. catch` controllers, because we are developing with a language that should use them according my memories :thinking: .  Why are we not using these keywords?

I know that to bring them can modify a lot of code, so you do not need to bring right now, but this is also something that we probably need to think. Of course I can be wrong, because I learned `c++` years ago and I did not work so much with it.

Best regards!

**Comment by vkalintiris** [comment]
> Hello @vkalintiris ,
> 
> I asked for some minor changes considering the codes we have for other collectors, but I have only one additional doubt: I am years without to develop with `C++`, but when I read our code I missed a lot `try .. catch` controllers, because we are developing with a language that should use them according my memories thinking . Why are we not using these keywords?
> 
> I know that to bring them can modify a lot of code, so you do not need to bring right now, but this is also something that we probably need to think. Of course I can be wrong, because I learned `c++` years ago and I did not work so much with it.
> 
> Best regards!

Practically, we should catch exceptions only for OOMs (ie. the possibility of getting any other exception is pretty much minimal and even then we wouldn't be able to handle in some sane way other than shutting down the agent).

Regarding OOMs, we are effectively allocating a small, fixed amount of memory for ML purposes, ie. other parts of the agent are far more likely to get an OOM error (eg. dbengine). Having said that I agree that we should add at some point a top-level exception handler for allocations.

**Comment by thiagoftsm** [comment]
Hello @andrewm4894 ,

I am starting test latest commits on this PR, please, test on your environment too.

Best regards!

**Review by thiagoftsm — APPROVED** [review]
After last commits, the PR looks like ready to be merged after hours testing, but I prefer to wait for @andrewm4894 's approve, to be sure we are not missing nothing.

**Review by MrZammler — APPROVED** [review]

**Comment by vkalintiris** [comment]
@andrewm4894 merging this because we only have 1 model by default. Ping me if you have any concerns.

**Comment by andrewm4894** [comment]
yep - looks good to me - ill be dropping and recreating nodes with various configs and in a few weeks will do a follow up PR with new defaults to take advantage of this PR.

**Pull request merged**